File: Spiral_Matrix.cpp
/*
Time Complexity: O(m*n)
Space Complexity: O(m*n)
Did this code successfully run on Leetcode :yes
Any problem you faced while coding this : no
*/
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int m = matrix.size(), n = matrix[0].size();
        int rs = 0, re = m-1, cs = 0, ce = n-1;
        
        while(ans.size()!=m*n){
            //right
            for(int i = cs; i<= ce; i++){
                ans.push_back(matrix[rs][i]);
            } rs++;
            //down
            for(int j = rs; j<=re; j++){
                ans.push_back(matrix[j][ce]);
            } ce--;
            //left
            if(rs<=re){
            for(int i = ce; i>=cs; i--){
                ans.push_back(matrix[re][i]);
            } re--;
            }
            //up
            if(cs<=ce){
            for(int j = re; j>=rs;j--){
                ans.push_back(matrix[j][cs]);
            } cs++;
            }
        }
        return ans;
    }
};

File: spiralMatrix.go
// Time Complexity : O(mn)
// Space Complexity : O(1) (returned array space is not considered)
// Did this code successfully run on Leetcode : Yes

/*

Approach:
In this problem, since we change direction in 4 ways, we keep a track of all
the boundaries using 4 variables - top, bottom, left, right.
We manipulate these variables in order to move in the directions we want (we
compute 4 directions per "spiral" loop, denoted by the 4 loops inside outer
loop).

*/

package main

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 {
		return []int{}
	}

	m := len(matrix)
	n := len(matrix[0])

	var res []int

	// use 4 pointers to move through the array
	t, b, l, r := 0, m-1, 0, n-1

	for t <= b && l <= r {
		// l -> r | top row
		for i := l; i <= r; i++ {
			res = append(res, matrix[t][i])
		}
		t++
		// t -> b | right column
		for i := t; i <= b; i++ {
			res = append(res, matrix[i][r])
		}
		r--
		// r -> l | bottom row
		if t <= b { // need this to check if all indices are in place
			for i := r; i >= l; i-- {
				res = append(res, matrix[b][i])
			}
			b--
		}
		// b -> t | left column
		if l <= r { // need this to check if all indices are in place
			for i := b; i >= t; i-- {
				res = append(res, matrix[i][l])
			}
			l++
		}
	}

	return res
}
